STM32 cmake refactoring, some test compiler warnings removed

db4ple opened the pull request:
- Reduced size/complexity of stm32/CMakeLists.txt
by combining some often used stm32 files in libraries and
source file list variables

- Aligned STM32 linker settings and compiler defines between
"traditional" Makefile and cmake generated Makefile. Result: smaller binaries generated by cmake stm32 builds

donmr commented:
I don't see the FREEDV_MODE_EN and CODEC2_MODE_EN settings in this, should they be there?
I have tried adding them but apparently don't know cmake well enough yet to make it work.

db4ple commented:
FREEDV_MODE_EN and CODEC2_MODE_EN: They are not there. However, I am thinking how to integrate these. We could have these always set (just have 700D and 1600 enabled for now). That is trivial.

db4ple commented:
Re Profiled Binaries: I carefully replicated the "original" use of profiled vs. non-profiled code. Most built binaries are not using profiled code.

db4ple commented:
In addition to my previous response: I now understood what you meant, the "new" unittest CMakeLists.txt compiled all binaries as profiled with semihosting. I fixed that. Please feel free to change the build type as required in a later pull request.

hobbes1069 commented:
It doesn't hurt to remove them but I left in some of the old Makefile stuff as a reminder that we need a replacement, either with a shell script or perhaps abuse CMake's custom target feature.

db4ple commented:
> 
> 
> It doesn't hurt to remove them but I left in some of the old Makefile stuff as a reminder that we need a replacement, either with a shell script or perhaps abuse CMake's custom target feature.

I am not sure what you are refering to?

hobbes1069 commented on stm32/CMakeLists.txt:
```diff
@@ -166,170 +166,131 @@ codebooknewamp1_energy.c
 # Rule for programming the SM1000
 #%.pgm: %.bin
 #	$(SUDO) dfu-util -d 0483:df11 -c 1 -i 0 -a 0 -s 0x08000000 -D $<
-
```
This section here, removing the old Makefile stuff.

hobbes1069 commented:
I get the following error when trying to build that I don't get on master:

```
[ 63%] Linking C executable sm1000.elf
/usr/bin/cmake -E cmake_link_script CMakeFiles/sm1000.dir/link.txt --verbose=1
/usr/lib64/ccache/arm-none-eabi-gcc -specs=nosys.specs -Wall -std=gnu11 -mlittle-endian -mthumb -mthumb-interwork --specs=nano.specs -mcpu=cortex-m4 -ffunction-sections -fdata-sections -fsingle-precision-constant -Wdouble-promotion -mfpu=fpv4-sp-d16 -mfloat-abi=hard -D__FPU_PRESENT=1 -D__FPU_USED=1 -g  -u_init -T/home/build/git/codec2/stm32/stm32_flash.ld -Xlinker --gc-sections CMakeFiles/sm1000.dir/src/sm1000_main.c.obj CMakeFiles/sm1000.dir/src/tone.c.obj CMakeFiles/sm1000.dir/src/sfx.c.obj CMakeFiles/sm1000.dir/src/sounds.c.obj CMakeFiles/sm1000.dir/src/morse.c.obj CMakeFiles/sm1000.dir/src/menu.c.obj CMakeFiles/sm1000.dir/src/tot.c.obj CMakeFiles/sm1000.dir/src/sm1000_leds_switches.c.obj CMakeFiles/sm1000.dir/home/build/git/codec2/src/codec2_fifo.c.obj CMakeFiles/sm1000.dir/src/debugblinky.c.obj CMakeFiles/sm1000.dir/src/stm32f4_vrom.c.obj CMakeFiles/sm1000.dir/home/build/git/codec2/src/lpc.c.obj CMakeFiles/sm1000.dir/home/build/git/codec2/src/nlp.c.obj CMakeFiles/sm1000.dir/home/build/git/codec2/src/postfilter.c.obj CMakeFiles/sm1000.dir/home/build/git/codec2/src/sine.c.obj CMakeFiles/sm1000.dir/home/build/git/codec2/src/codec2.c.obj CMakeFiles/sm1000.dir/home/build/git/codec2/src/codec2_fft.c.obj CMakeFiles/sm1000.dir/home/build/git/codec2/src/gp_interleaver.c.obj CMakeFiles/sm1000.dir/home/build/git/codec2/src/interldpc.c.obj CMakeFiles/sm1000.dir/home/build/git/codec2/src/kiss_fft.c.obj CMakeFiles/sm1000.dir/home/build/git/codec2/src/kiss_fftr.c.obj CMakeFiles/sm1000.dir/home/build/git/codec2/src/interp.c.obj CMakeFiles/sm1000.dir/home/build/git/codec2/src/lsp.c.obj CMakeFiles/sm1000.dir/home/build/git/codec2/src/ofdm.c.obj CMakeFiles/sm1000.dir/home/build/git/codec2/src/phase.c.obj CMakeFiles/sm1000.dir/home/build/git/codec2/src/quantise.c.obj CMakeFiles/sm1000.dir/home/build/git/codec2/src/pack.c.obj CMakeFiles/sm1000.dir/home/build/git/codec2/src/dump.c.obj CMakeFiles/sm1000.dir/home/build/git/codec2/src/fdmdv.c.obj CMakeFiles/sm1000.dir/home/build/git/codec2/src/freedv_api.c.obj CMakeFiles/sm1000.dir/home/build/git/codec2/src/filter.c.obj CMakeFiles/sm1000.dir/home/build/git/codec2/src/varicode.c.obj CMakeFiles/sm1000.dir/home/build/git/codec2/src/golay23.c.obj CMakeFiles/sm1000.dir/home/build/git/codec2/src/fsk.c.obj CMakeFiles/sm1000.dir/home/build/git/codec2/src/fmfsk.c.obj CMakeFiles/sm1000.dir/home/build/git/codec2/src/freedv_vhf_framing.c.obj CMakeFiles/sm1000.dir/home/build/git/codec2/src/freedv_data_channel.c.obj CMakeFiles/sm1000.dir/home/build/git/codec2/src/newamp1.c.obj CMakeFiles/sm1000.dir/home/build/git/codec2/src/mbest.c.obj CMakeFiles/sm1000.dir/home/build/git/codec2/src/cohpsk.c.obj CMakeFiles/sm1000.dir/home/build/git/codec2/src/HRA_112_112.c.obj CMakeFiles/sm1000.dir/home/build/git/codec2/src/linreg.c.obj CMakeFiles/sm1000.dir/home/build/git/codec2/src/mpdecode_core.c.obj CMakeFiles/sm1000.dir/home/build/git/codec2/src/phi0.c.obj CMakeFiles/sm1000.dir/home/build/git/codec2/src/HRAb_396_504.c.obj CMakeFiles/sm1000.dir/codebook.c.obj CMakeFiles/sm1000.dir/codebookd.c.obj CMakeFiles/sm1000.dir/codebookjvm.c.obj CMakeFiles/sm1000.dir/codebookge.c.obj CMakeFiles/sm1000.dir/codebookmel.c.obj CMakeFiles/sm1000.dir/codebooknewamp1_energy.c.obj CMakeFiles/sm1000.dir/codebooknewamp1.c.obj CMakeFiles/sm1000.dir/src/system_stm32f4xx.c.obj CMakeFiles/sm1000.dir/src/startup_stm32f4xx.s.obj CMakeFiles/sm1000.dir/src/gdb_stdio.c.obj CMakeFiles/sm1000.dir/src/init.c.obj  -o sm1000.elf -lg -lm libstm32f4_adac.a libstm32f4.a libCMSIS.a -lg -lm
CMakeFiles/sm1000.dir/src/init.c.obj: In function `_init':
/home/build/git/codec2/stm32/src/init.c:6: multiple definition of `_init'
/usr/lib/gcc/arm-none-eabi/7.4.0/thumb/v7e-m/fpv4-sp/hard/crti.o:(.init+0x0): first defined here
CMakeFiles/sm1000.dir/src/init.c.obj: In function `_fini':
init.c:(.text._fini+0x0): multiple definition of `_fini'
/usr/lib/gcc/arm-none-eabi/7.4.0/thumb/v7e-m/fpv4-sp/hard/crti.o:(.fini+0x0): first defined here
collect2: error: ld returned 1 exit status
```

db4ple commented:
I just fixed that by force-pushing some changes. Please update your local branch by replacing it with the "right" content from my repo.

hobbes1069 commented:
Ok, working on the kinks, I'm still not used to being on this side of github :) Usually I'm the guy submitting a pull request.

hobbes1069 commented:
That seemed to fix it for me... @donmr?

drowe67 commented:
Has anyone actually flashed a SM1000 and tested the binary?  (I haven't for a while).

donmr commented:
I just merged your branch into my workspace and I see the _init errors.

[100%] Linking C executable tst_api_demod.elf /usr/bin/../lib/gcc/arm-none-eabi/7.3.1/../../../../arm-none-eabi/lib/thumb/v7e-m/fpv4-sp/hard/libg_nano.a(lib_a-init.o): In function `__libc_init_array':
init.c:(.text.__libc_init_array+0x14): undefined reference to `_init'
collect2: error: ld returned 1 exit status
...

donmr commented:
I did it with the 1600 version from SVN before we started all of this
Git and CMake stuff.  That seemed to work fine.

Don

On Mon, Apr 08, 2019 at 02:24:06PM -0700, drowe67 wrote:
> Has anyone actually flashed a SM1000 and tested the binary?  (I haven't for a while).
> 
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/drowe67/codec2/pull/19#issuecomment-481013670

db4ple commented:
@donmr : I am not sure why you are getting `_init undefined` errors. I, @hobbes1069 and also the Travis build have no problems with that. It might be a good idea to try building the stm32 build from scratch (i.e. empty build_stm32 dir and running cmake) to see if the issue persists.

donmr commented:
I was missing some other commits from master.  After updating it works fine.

drowe67 commented:
> I was missing some other commits from master. After updating it works fine.

That's great Don - is the memory usage issue you reported a few days OK now OK?

db4ple commented:
While I don't exactly know how @donmr creates the .data statistics shown in #16, I tested the freedv_rx_profile build and the map shows for instance nothing is being used from cohpsk.c 
The total ram allocation (.data + .bss) is  16f0 (5872 bytes), .data being just 1d8 (472 bytes) according to 
the map file (I added some extra commands to cmake to generate a map file, which unfortunately generates the map file for all binaries with the same name, have to figure out how to get the target executable name in cmake).

db4ple commented:
Maybe a simpler way to check this without too much detail is to use the command shown below. You clearly see static RAM usage is defined by .data + .bss + ._user_heap_stack, everything else is available for malloc.
```
~/gcc-arm-none-eabi-8-2018-q4-major/bin/arm-none-eabi-size --format=SysV freedv_rx_profile.elf

freedv_rx_profile.elf  :
section               size         addr
.isr_vector            392    134217728
.eeprom              49152    134234112
.bkpsram                 0   1073889280
.text               306336    134283264
.init_array              4    134589600
.fini_array              4    134589604
.data                  476    536870912
.bss                  4372    536871388
._user_heap_stack     1024    536875760
.ARM.attributes         48            0
.comment               127            0
.debug_info         195544            0
.debug_abbrev        21360            0
.debug_aranges        3920            0
.debug_ranges         3216            0
.debug_line          50690            0
.debug_str           22332            0
.debug_frame         21220            0
.stab                  204            0
.stabstr               493            0
Total               680914
```

donmr commented:
I had the linker produce a map file (target_link_libraries(tst_api_demod ${SEMIHOSTING_LIBS} -Wl,-Map=tst_api_demod.map)) and then wrote a small python program to scan it.  It sums up the sizes of the entries for RAM::.data.   

On freedv_rx_profile I get:

RAM::.data 476
    168 libcodec2_prof.a(gp_interleaver.c.obj)
    144 libcodec2_prof.a(fdmdv.c.obj)
    100 /usr/bin/../lib/gcc/arm-none-eabi/7.3.1/../../../../arm-none-eabi/lib/thumb/v7e-m/fpv4-sp/hard/libg_nano.a(lib_a-impure.o)
    32 libcodec2_prof.a(mpdecode_core.c.obj)
    12 CMakeFiles/freedv_rx_profile.dir/src/stm32f4_machdep.c.obj
    12 libcodec2_prof.a(freedv_api.c.obj)
    4 libcodec2_prof.a(sine.c.obj)
    1 /usr/bin/../lib/gcc/arm-none-eabi/7.3.1/../../../../arm-none-eabi/lib/thumb/v7e-m/fpv4-sp/hard/libm.a(lib_a-s_lib_ver.o)

Is the compiler (linker, etc.) smart enough to see that freedv_open is only called with the 1600 mode and leave out everything else??  My tst_api tests take the mode as a run time configuration so the will have all of the code in them.  That is why I need the make time config stuff.

Thanks for the pointer to size(1), I have not encountered that tool before.

db4ple commented:
If you compile with -O1 or better compiler should be able to detect this w/o problems. That is why I added the FREEDV_MODE_ACTIVE macro stuff, which helps us to explicitly tell the compiler that that a given mode should not be supported at compile time. With this, optimization will be able to kick in (if we optimize the code) and remove the respective if {...} block statements calling some maybe some other functions. The linker will then do the rest for us. Especially since cmake uses the splitting of and .o file into smaller, independent sections per function/variable now (I added this from your unittest Makefile settings).

With this, only truly needed functions / variables are included instead of the whole .o contents even if just a single function is referenced from that .o file.

BTW, I just discovered the same trick as you to produce maps for executables in cmake. I will add a macro 
add_mapped_executable to cmake, which essentially automates the process of generating maps.

drowe67 commented:
@db4ple @donmr is this ready for merge?

db4ple commented:
> @db4ple @donmr is this ready for merge?
Yes, from my point of view. I try to make pull request only when I think something is ready.

donmr commented:
Yes merge it.  If anything more is needed we can work from there.

drowe67 commented:
> > @db4ple @donmr is this ready for merge?
> > Yes, from my point of view. I try to make pull request only when I think something is ready.

I'm also in the habit of making  PRs for work in progress, e.g. to gather comments early, or collaborate with others.  Pls nudge me when a PR is ready for merge.
